core/types/state_account_storage.libevm_test.go

```diff
@@ -0,0 +1,153 @@
+		assertExtra         func(*testing.T, *types.StateAccount)
+		wantTrieHash        common.Hash
+	}{
+		{
+			name: "vanilla geth",
+			registerAndSetExtra: func(a *types.StateAccount) *types.StateAccount {
+				return a
+			},
+			assertExtra: func(t *testing.T, a *types.StateAccount) {
+				t.Helper()
+				assert.Truef(t, a.Extra.Equal(nil), "%T.%T.IsEmpty()", a, a.Extra)
+			},
+			wantTrieHash: vanillaGeth,
+		},
+		{
+			name: "true-boolean payload",
+			registerAndSetExtra: func(a *types.StateAccount) *types.StateAccount {
+				types.RegisterExtras[bool]().SetOnStateAccount(a, true)
+				return a
+			},
+			assertExtra: func(t *testing.T, sa *types.StateAccount) {
+				t.Helper()
+				assert.Truef(t, types.ExtraPayloads[bool]{}.FromStateAccount(sa), "")
+			},
+			wantTrieHash: trueBool,
+		},
+		{
+			name: "explicit false-boolean payload",
+			registerAndSetExtra: func(a *types.StateAccount) *types.StateAccount {
+				p := types.RegisterExtras[bool]()
+				p.SetOnStateAccount(a, false) // the explicit part
+				return a
+			},
+			assertExtra: func(t *testing.T, sa *types.StateAccount) {
+				t.Helper()
+				assert.Falsef(t, types.ExtraPayloads[bool]{}.FromStateAccount(sa), "")
+			},
+			wantTrieHash: falseBool,
+		},
+		{
+			name: "implicit false-boolean payload",
+			registerAndSetExtra: func(a *types.StateAccount) *types.StateAccount {
+				types.RegisterExtras[bool]()
+				// Note that `a` is reflected, unchanged (the implicit part).
+				return a
+			},
+			assertExtra: func(t *testing.T, sa *types.StateAccount) {
+				t.Helper()
+				assert.Falsef(t, types.ExtraPayloads[bool]{}.FromStateAccount(sa), "")
+			},
+			wantTrieHash: falseBool,
+		},
+		{
+			name: "arbitrary payload",
+			registerAndSetExtra: func(a *types.StateAccount) *types.StateAccount {
+				p := arbitraryPayload{arbitraryData}
+				types.RegisterExtras[arbitraryPayload]().SetOnStateAccount(a, p)
+				return a
+			},
+			assertExtra: func(t *testing.T, sa *types.StateAccount) {
+				t.Helper()
+				got := types.ExtraPayloads[arbitraryPayload]{}.FromStateAccount(sa)
+				assert.Equalf(t, arbitraryPayload{arbitraryData}, got, "")
+			},
+			wantTrieHash: arbitrary,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			types.TestOnlyClearRegisteredExtras()
+			t.Cleanup(types.TestOnlyClearRegisteredExtras)
+
+			acct := tt.registerAndSetExtra(&types.StateAccount{
+				Nonce:    42,
+				Balance:  uint256.NewInt(314159),
+				Root:     types.EmptyRootHash,
+				CodeHash: types.EmptyCodeHash[:],
+			})
+
+			db := triedb.NewDatabase(rawdb.NewMemoryDatabase(), nil)
+			id := trie.TrieID(types.EmptyRootHash)
+			state, err := trie.NewStateTrie(id, db)
+			require.NoError(t, err, "trie.NewStateTrie(types.EmptyRootHash, ...)")
+
+			require.NoErrorf(t, state.UpdateAccount(addr, acct), "%T.UpdateAccount(...)", state)
+			assert.Equalf(t, tt.wantTrieHash, state.Hash(), "%T.Hash() after UpdateAccount()", state)
```

Comment:
Can can we also add testing for Copy & Snapshot/Revert being handled properly?
For example statedb.Copy makes a shallow copy of the accounts modified, so we should ensure that setting the extras on the origin does not impact the copied statedb & the other way around.

Reply:
Snapshot/Revert covered in #48 

I'm about to test Copy too. If it fails then I'll see how much needs to change and might start a new PR if the fix spreads too far.